core/src/main/java/io/grpc/internal/SerializingExecutor.java

```diff
@@ -82,90 +69,48 @@ public SerializingExecutor(Executor executor) {
-      }
-    }
-    if (scheduleTaskRunner) {
-      boolean threw = true;
+  private void schedule() {
+    if (running.compareAndSet(false, true)) {
+      boolean success = false;
       try {
-        executor.execute(taskRunner);
-        threw = false;
+        executor.execute(this);
+        success = true;
       } finally {
-        if (threw) {
-          synchronized (internalLock) {
-            // It is possible that at this point that there are still tasks in
-            // the queue, it would be nice to keep trying but the error may not
-            // be recoverable.  So we update our state and propogate so that if
-            // our caller deems it recoverable we won't be stuck.
-            isThreadScheduled = false;
-          }
+        // It is possible that at this point that there are still tasks in
+        // the queue, it would be nice to keep trying but the error may not
+        // be recoverable.  So we update our state and propogate so that if
+        // our caller deems it recoverable we won't be stuck.
+        if (!success) {
+          running.set(false);
         }
       }
     }
   }
 
-  /**
-   * Task that actually runs the Runnables.  It takes the Runnables off of the
-   * queue one by one and runs them.  After it is done with all Runnables and
-   * there are no more to run, puts the SerializingExecutor in the state where
-   * isThreadScheduled = false and returns.  This allows the current worker
-   * thread to return to the original pool.
-   */
-  private class TaskRunner implements Runnable {
-    @Override
-    public void run() {
-      boolean stillRunning = true;
+  @Override
+  public void run() {
+    Runnable r;
+    while ((r = runQueue.poll()) != null) {
       try {
-        while (true) {
-          Runnable nextToRun;
-          synchronized (internalLock) {
-            Preconditions.checkState(isThreadScheduled);
-            nextToRun = waitQueue.poll();
-            if (nextToRun == null) {
-              isThreadScheduled = false;
-              stillRunning = false;
-              break;
-            }
-          }
-
-          // Always run while not holding the lock, to avoid deadlocks.
-          try {
-            nextToRun.run();
-          } catch (RuntimeException e) {
-            // Log it and keep going.
-            log.log(Level.SEVERE, "Exception while executing runnable "
-                + nextToRun, e);
-          }
-        }
-      } finally {
-        if (stillRunning) {
-          // An Error is bubbling up, we should mark ourselves as no longer
-          // running, that way if anyone tries to keep using us we won't be
-          // corrupted.
-          synchronized (internalLock) {
-            isThreadScheduled = false;
-          }
-        }
+        r.run();
+      } catch (RuntimeException e) {
+        // Log it and keep going.
+        log.log(Level.SEVERE, "Exception while executing runnable " + r, e);
       }
     }
+    running.set(false);
```

Comment:
This isn't protected against exceptions other than RuntimeException. It seems we would want it to deal with others.